builtin/gc.c

```diff
@@ -1345,6 +1345,154 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+
+static void copy_file(const char *srcdir, const char *dstdir, const char *name)
+{
+	int ret = 0;
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+	int srcfd = -1, dstfd = -1;
+	char buf[1024];
+	ssize_t nr;
+
+	strbuf_addf(&src, "%s/%s", srcdir, name);
+	strbuf_addf(&dst, "%s/%s", dstdir, name);
+
+	srcfd = open(src.buf, O_RDONLY);
+	if (srcfd < 0) {
+		error("failed to open source file");
+		ret = 1;
+		goto cleanup;
+	}
+
+	dstfd = open(dst.buf, O_WRONLY | O_CREAT | O_TRUNC, 0644);
+	if (dstfd < 0) {
+		error("failed to open destination file");
+		ret = 1;
+		goto cleanup;
+	}
+
+	while ((nr = read(srcfd, buf, sizeof(buf))) > 0) {
+		if (write(dstfd, buf, nr) < 0) {
+			error("failed to write to destination file");
+			ret = 1;
+		}
+	}
+
+cleanup:
+	if (srcfd >= 0) close(srcfd);
+	if (dstfd >= 0) close(dstfd);
+
+	if (ret)
+		die_errno(_("failed to copy '%s' to '%s'"), src.buf, dst.buf);
+
+	strbuf_release(&src);
+	strbuf_release(&dst);
+}
+
+static void move_file(const char *srcdir, const char *dstdir, const char *name)
+{
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+
+	strbuf_addf(&src, "%s/%s", srcdir, name);
+	strbuf_addf(&dst, "%s/%s", dstdir, name);
+
+	if (rename(src.buf, dst.buf))
+		die_errno(_("failed to move '%s' to '%s'"), src.buf, dst.buf);
+
+	strbuf_release(&src);
+	strbuf_release(&dst);
+}
+
+static void delete_file(const char *dir, const char *name)
+{
+	struct strbuf path = STRBUF_INIT;
+
+	strbuf_addf(&path, "%s/%s", dir, name);
+
+	if (unlink(path.buf))
+		warning_errno(_("failed to delete '%s'"), path.buf);
+
+	strbuf_release(&path);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct strbuf path = STRBUF_INIT;
+	struct stat st;
+	char *basename, *keep_filename, *rev_filename, *idx_filename;
+	int has_keep, has_rev;
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+	keep_filename = xstrfmt("%s.keep", basename);
+	rev_filename = xstrfmt("%s.rev", basename);
+	idx_filename = xstrfmt("%s.idx", basename);
```

Comment:
What should we do with `.pack` files that have no associated `.idx` file?

(Could be an in-progress `git fetch`? Although that would probably have a `tmp_*` file at first, there is a small time window when it is already renamed to `.pack` but the `.idx` has not been renamed yet; But it could also be a left-over from some failed operation...).

Reply:
We now skip if the `.idx` file is missing, meaning we need both the `.idx` and the `.pack` to start migration. If we just miss the `.idx` creation, then we'll catch it next time the maintenance is run.